### PR TITLE
[clang-analysis]Fix false positive in mutation check when using pointer to member function

### LINKED ISSUE #66204: false positive for clang-tidy rule 'misc-const-correctness'
I use clang-tidy 15.0.7 from Fedora 37

consider the following reproducer
```
#include <iostream>

class X {
public:
	int a;
	int b;
};

void func( X& x, int X::* m )
{
	/*const*/ X& tmp = x;

	tmp.*m = 77;
}

int main()
{
	X x;
	x.a = 0;
	x.b = 0;

	std::cout << x.a << " " << x.b << std::endl;

	func(x, &X::a);

	std::cout << x.a << " " << x.b << std::endl;

	return 0;
}
```

The ```/* const */``` marked line gets tagged with 'misc-const-correctness'.
Though adding that 'const' clearly results in a compilation error with g++ (12.3.1) or clang++ (15.0.7)

### PR DESCRIPTION
**Pull request by HerrCai0907**
Fixes: #66204

### CODE REVIEW

**Review by PiotrZSL — APPROVED** [review]
Clang-tidy part looking fine for me.

File: clang/lib/Analysis/ExprMutationAnalyzer.cpp [hunk 1]
```diff
@@ -100,6 +100,20 @@ AST_MATCHER(CXXTypeidExpr, isPotentiallyEvaluated) {
   return Node.isPotentiallyEvaluated();
 }
 
+AST_MATCHER(CXXMemberCallExpr, isConstCallee) {
+  const auto *CalleeDecl = Node.getCalleeDecl();
```
Comment (EugeneZelenko) [inline clang/lib/Analysis/ExprMutationAnalyzer.cpp:104]: Please don't use `auto` unless type is explicitly stated in same statement or iterator.

File: clang/lib/Analysis/ExprMutationAnalyzer.cpp [hunk 2]
```diff
@@ -100,6 +100,20 @@ AST_MATCHER(CXXTypeidExpr, isPotentiallyEvaluated) {
...
+  const auto *VD = dyn_cast_or_null<ValueDecl>(CalleeDecl);
+  if (!VD)
+    return false;
+  const auto T = VD->getType().getCanonicalType();
```
Comment (EugeneZelenko) [inline clang/lib/Analysis/ExprMutationAnalyzer.cpp:108]: Ditto.

File: clang/lib/Analysis/ExprMutationAnalyzer.cpp [hunk 3]
```diff
@@ -391,7 +405,9 @@ const Stmt *ExprMutationAnalyzer::findMemberMutation(const Expr *Exp) {
       match(findAll(expr(anyOf(memberExpr(hasObjectExpression(
                                    canResolveToExpr(equalsNode(Exp)))),
                                cxxDependentScopeMemberExpr(hasObjectExpression(
-                                   canResolveToExpr(equalsNode(Exp))))))
+                                   canResolveToExpr(equalsNode(Exp)))),
+                               binaryOperator(allOf(hasOperatorName(".*"),
```
Comment (mzyKi) [inline clang/lib/Analysis/ExprMutationAnalyzer.cpp:409]: ```allOf``` is redundant.I think ```binaryOperator(hasOperatorName(".*"),hasLHS(equalsNode(Exp))``` is OK.

File: clang/lib/Analysis/ExprMutationAnalyzer.cpp [hunk 4]
```diff
@@ -274,8 +288,8 @@ const Stmt *ExprMutationAnalyzer::findDirectMutation(const Expr *Exp) {
   const auto NonConstMethod = cxxMethodDecl(unless(isConst()));
 
   const auto AsNonConstThis = expr(anyOf(
-      cxxMemberCallExpr(callee(NonConstMethod),
-                        on(canResolveToExpr(equalsNode(Exp)))),
+      cxxMemberCallExpr(on(canResolveToExpr(equalsNode(Exp))),
+                        unless(isConstCallee())),
```
Comment (mzyKi) [inline clang/lib/Analysis/ExprMutationAnalyzer.cpp:292]: ```unless(isConstCallee())``` This condition is unnecessary. I have tested your code in my machine.
In such testcase example,
```cpp
#include <iostream>

class X {
public:
  int a;
  int b;
};

void func(X &x, int X::*m) {
  /*const*/ X &tmp = x;

  tmp.*m = 77;
}

int main() {
  X x;
  x.a = 0;
  x.b = 0;

  std::cout << x.a << " " << x.b << std::endl;

  func(x, &X::a);

  std::cout << x.a << " " << x.b << std::endl;

  return 0;
}
```
 use ```cxxMemberCallExpr(callee(NonConstMethod),
                        on(canResolveToExpr(equalsNode(Exp)))),``` cxxMemberCallExpr matcher do not produce false positive

Comment (HerrCai0907) [reply clang/lib/Analysis/ExprMutationAnalyzer.cpp:292]: see this test case

```c++
buildASTFromCode("struct X {};"
                 "using T = int (X::*)() const;"
                 "void f(X &x, T m) { X &ref = x; (ref.*m)(); }");
```

**Pull request merged**